### PR TITLE
advanced setting UI construction

### PR DESCRIPTION
**Pull request by DhaaraniCIT**

### CODE REVIEW

**Comment by DhaaraniCIT** [comment]
UI
Screenshot 2023-11-15 at 2 15 03 PM
Screenshot 2023-11-15 at 2 15 20 PM
Screenshot 2023-11-15 at 2 15 37 PM

**Comment by DhaaraniCIT** [comment]
fixed the PR comments of https://github.com/fylein/fyle-integrations-settings-app/pull/254 and https://github.com/fylein/fyle-integrations-settings-app/pull/259 in this PR

File: src/app/core/models/sage300/sage300-configuration/sage300-advanced-settings.mode.ts [hunk 1]
```diff
@@ -124,26 +53,25 @@ export class Sage300AdvancedSettingModel {
             autoCreateMerchantDestinationEntity: new FormControl(advancedSettings?.auto_create_merchant_destination_entity ? true : false),
             syncSage300ToFylePayments: new FormControl(advancedSettings?.sync_sage_300_to_fyle_payments ? true : false),
             autoCreateDestinationEntity: new FormControl(advancedSettings?.auto_create_destination_entity ? true : false),
-            skipExport: new FormControl(advancedSettings?.skipExport ? true : false),
             memoStructure: new FormControl(advancedSettings?.memo_structure ? advancedSettings?.memo_structure : null ),
             defaultJobName: new FormControl(advancedSettings?.default_job_name ? advancedSettings?.default_job_name : null ),
```
Comment (ashwin1111) [inline src/app/core/models/sage300/sage300-configuration/sage300-advanced-settings.mode.ts:57]: we cannot assign this directly, check how it's done in intacct

File: src/app/core/models/sage300/sage300-configuration/sage300-advanced-settings.mode.ts [hunk 2]
```diff
@@ -18,8 +19,7 @@ export type Sage300AdvancedSetting = {
     schedule_enabled: boolean,
     emails_selected: EmailOption[],
     emails_added: EmailOption[],
-    auto_map_vendor: boolean,
-    skipExport: boolean
+    auto_map_vendor: boolean
```
Comment (ashwin1111) [inline src/app/core/models/sage300/sage300-configuration/sage300-advanced-settings.mode.ts:22]: `auto_create_vendor`

File: src/app/core/models/sage300/sage300-configuration/sage300-advanced-settings.mode.ts [hunk 3]
```diff
@@ -124,26 +53,25 @@ export class Sage300AdvancedSettingModel {
...
-            skipExport: new FormControl(advancedSettings?.skipExport ? true : false),
             memoStructure: new FormControl(advancedSettings?.memo_structure ? advancedSettings?.memo_structure : null ),
             defaultJobName: new FormControl(advancedSettings?.default_job_name ? advancedSettings?.default_job_name : null ),
-            defaultJobId: new FormControl(advancedSettings?.default_job_id ? advancedSettings?.default_job_id : null ),
             scheduleEnabled: new FormControl(advancedSettings?.schedule_enabled ? true : false),
-            emailsSelected: new FormControl(advancedSettings?.emails_selected ? advancedSettings?.emails_selected : null),
+            email: new FormControl(advancedSettings?.emails_selected ? advancedSettings?.emails_selected : []),
             emailsAdded: new FormControl(advancedSettings?.emails_added ? advancedSettings?.emails_added : null),
-            autoMapVendor: new FormControl(advancedSettings?.auto_map_vendor ? true : false)
+            autoMapVendor: new FormControl(advancedSettings?.auto_map_vendor ? true : false),
```
Comment (ashwin1111) [inline src/app/core/models/sage300/sage300-configuration/sage300-advanced-settings.mode.ts:61]: autoCreateVendor

File: src/app/core/models/sage300/sage300-configuration/sage300-advanced-settings.mode.ts [hunk 4]
```diff
@@ -124,26 +53,25 @@ export class Sage300AdvancedSettingModel {
...
+            scheduleAutoExportFrequency: new FormControl(1),
```
Comment (ashwin1111) [inline src/app/core/models/sage300/sage300-configuration/sage300-advanced-settings.mode.ts:62]: why are we hard coding?

File: src/app/core/models/sage300/sage300-configuration/sage300-advanced-settings.mode.ts [hunk 5]
```diff
@@ -124,26 +53,25 @@ export class Sage300AdvancedSettingModel {
...
+            topLevelMemo: new FormControl(null)
```
Comment (ashwin1111) [inline src/app/core/models/sage300/sage300-configuration/sage300-advanced-settings.mode.ts:63]: do don't have 2 memo

File: src/app/core/models/sage300/sage300-configuration/sage300-advanced-settings.mode.ts [hunk 6]
```diff
@@ -1,5 +1,6 @@
 import { FormControl, FormGroup } from "@angular/forms";
```
Comment (ashwin1111) [inline src/app/core/models/sage300/sage300-configuration/sage300-advanced-settings.mode.ts:1]: file name typo, .model.ts

File: src/app/integrations/sage300/sage300-shared/sage300-advanced-settings/sage300-advanced-settings.component.html [hunk 7]
```diff
@@ -1 +1,176 @@
...
+                        You can automate the export and sync of your data in this section.
+                    </h5>
+                </div>
+            </div>
+            <div class="tw-mb-16-px">
+                <div class="tw-rounded-12-px tw-border-separator tw-border tw-bg-white tw-mb-16-px">
+                    <app-configuration-toggle-field 
+                        [form]="advancedSettingForm"
+                        [label]="'Schedule automatic export'" 
+                        [subLabel]="'Set up a schedule to frequently automate the export of expenses from Fyle to QuickBooks Desktop.'"
```
Comment (ashwin1111) [inline src/app/integrations/sage300/sage300-shared/sage300-advanced-settings/sage300-advanced-settings.component.html:32]: app name says qbd

File: src/app/integrations/sage300/sage300-shared/sage300-advanced-settings/sage300-advanced-settings.component.html [hunk 8]
```diff
@@ -1 +1,176 @@
...
+                            [mandatoryErrorListName]="'email for error notification'"
+                            [subLabel]="'Select or add an email address to notify upon import failure or integration error.'"
+                            [options]="adminEmails"
+                            [placeholder]="'Select email address'"
+                            [formControllerName]="'email'">
+                        </app-email-multi-select-field> -->
+                    </div>
+                </div>
+                <div *ngIf="isReimbursableExpense" class="tw-rounded-12-px tw-border-separator tw-border tw-bg-white">
+                    <app-configuration-toggle-field
```
Comment (ashwin1111) [inline src/app/integrations/sage300/sage300-shared/sage300-advanced-settings/sage300-advanced-settings.component.html:61]: comment this out too, we'll do in v2 only - payment sync

File: src/app/integrations/sage300/sage300-shared/sage300-advanced-settings/sage300-advanced-settings.component.html [hunk 9]
```diff
@@ -1 +1,176 @@
...
+                    <h5 class="tw-text-faded-text-color tw-text-14-px tw-pt-8-px !tw-font-400">
+                        In this section, you can customize the data that you'd like to export from Fyle to Sage 300 CRE. You can choose what data points need to be exported and what shouldn't be.
+                    </h5>
+                </div>
+            </div>
+            <div class="tw-rounded-12-px tw-border-separator tw-border tw-bg-white tw-mb-16-px">
+                <app-configuration-select-field
+                    [form]="advancedSettingForm"
+                    [isFieldMandatory]="false"
+                    [mandatoryErrorListName]="'Top level description'"
```
Comment (ashwin1111) [inline src/app/integrations/sage300/sage300-shared/sage300-advanced-settings/sage300-advanced-settings.component.html:98]: only 1 memo

File: src/app/integrations/sage300/sage300-shared/sage300-advanced-settings/sage300-advanced-settings.component.html [hunk 10]
```diff
@@ -1 +1,176 @@
...
+                </div>
+            </div>
+            <div class="tw-rounded-12-px tw-border-separator tw-border tw-bg-white tw-mb-16-px">
+                <app-configuration-toggle-field 
+                    [form]="advancedSettingForm"
+                    [label]="'Skip export of specific expenses from Fyle to Sage 300 CRE'" 
+                    [subLabel]="'You could choose to skip expenses from Fyle to Sage 300 CRE by setting up a conditional rule.'" 
+                    [formControllerName]="'skipExport'"
+                    [iconPath]="'expense'"
+                    [redirectLink]="'jhjh'">
```
Comment (ashwin1111) [inline src/app/integrations/sage300/sage300-shared/sage300-advanced-settings/sage300-advanced-settings.component.html:140]: read link from enum

File: src/app/integrations/sage300/sage300-shared/sage300-advanced-settings/sage300-advanced-settings.component.html [hunk 11]
```diff
@@ -1 +1,176 @@
...
+                        Based on your preference, you can choose whether you want to create any new records in Sage 300 CRE from Fyle. (when there is no employee record found, or when the accounting period is closed)
+                    </h5>
+                </div>
+            </div>
+            <div class="tw-rounded-12-px tw-border-separator tw-border tw-bg-white">
+                <app-configuration-toggle-field 
+                    [form]="advancedSettingForm"
+                    [label]="'Auto-Create Vendor'" 
+                    [subLabel]="'When matching Vendor record is not found for an employee in Fyle, the integration will auto-create a new record in Sage 300 CRE.'" 
+                    [formControllerName]="'autoCreateMerchantDestinationEntity'"
```
Comment (ashwin1111) [inline src/app/integrations/sage300/sage300-shared/sage300-advanced-settings/sage300-advanced-settings.component.html:167]: controller name might need to change, autoCreateVendor

File: src/app/integrations/sage300/sage300-shared/sage300-advanced-settings/sage300-advanced-settings.component.ts [hunk 12]
```diff
@@ -28,11 +33,66 @@ export class Sage300AdvancedSettingsComponent implements OnInit {
...
+  hours: HourOption[] = [];
+
+  isReimbursableExpense: boolean = false;
+
+  ConfigurationCtaText = ConfigurationCta;
+
+  defaultMemoOptions: string[] = ['employee_email', 'purpose', 'category', 'spent_on', 'report_number', 'expense_link'];
+
+  defaultTopMemoOptions: string[] = ['employee_email', 'purpose', 'category', 'spent_on', 'report_number', 'expense_link'];
+
```
Comment (ashwin1111) [inline src/app/integrations/sage300/sage300-shared/sage300-advanced-settings/sage300-advanced-settings.component.ts:51]: 1 memo

File: src/app/integrations/sage300/sage300-shared/sage300-advanced-settings/sage300-advanced-settings.component.ts [hunk 13]
```diff
@@ -143,18 +214,23 @@ export class Sage300AdvancedSettingsComponent implements OnInit {
...
 
   private getSettingsAndSetupForm(): void {
+    for (let i = 1; i <= 24; i++) {
+      this.hours.push({ label: `${i}`, value: i });
+    }
     forkJoin([
       this.advancedSettingsService.getAdvancedSettings().pipe(catchError(() => of(null))),
-      this.advancedSettingsService.getAdminEmail(),
+      this.exportSettingsService.getSage300ExportSettings(),
+      // This.advancedSettingsService.getAdminEmail(),
```
Comment (ashwin1111) [inline src/app/integrations/sage300/sage300-shared/sage300-advanced-settings/sage300-advanced-settings.component.ts:223]: can remove this line completely, we'll add later when required

File: src/app/integrations/sage300/sage300-shared/sage300-advanced-settings/sage300-advanced-settings.component.ts [hunk 14]
```diff
@@ -143,18 +214,23 @@ export class Sage300AdvancedSettingsComponent implements OnInit {
...
       this.advancedSettingsService.getExpenseFilter(),
       this.advancedSettingsService.getExpenseFilelds()
-    ]).subscribe((responses) => {
-      this.advancedSetting = responses[0];
-      this.adminEmails = responses[1];
-      this.expenseFilters = responses[2].results;
-      this.conditionFieldOptions = responses[3];
+    ]).subscribe(([sage300AdvancedSettingResponse, exportSettingsResponse, expenseFiltersGet, ExpenseFilterCondition]) => {
+      this.advancedSetting = sage300AdvancedSettingResponse;
+      // This.adminEmails = adminEmails;
```
Comment (ashwin1111) [inline src/app/integrations/sage300/sage300-shared/sage300-advanced-settings/sage300-advanced-settings.component.ts:228]: can remove this line completely, we'll add later when required

File: src/app/shared/components/configuration/configuration-multi-select/configuration-multi-select.component.html [hunk 15]
```diff
@@ -1,4 +1,4 @@
-<div [formGroup]="form" class="tw-p-24-px tw-flex tw-flex-row tw-space-between">
+<div [formGroup]="form" class="tw-p-24-px tw-flex tw-justify-between tw-items-center">
```
Comment (ashwin1111) [inline src/app/shared/components/configuration/configuration-multi-select/configuration-multi-select.component.html:1]: please test intacct and other places which uses this code

File: src/app/shared/components/configuration/configuration-select-field/configuration-select-field.component.html [hunk 16]
```diff
@@ -1,6 +1,6 @@
-<div [formGroup]="form" class="tw-p-24-px tw-flex tw-flex-row tw-space-between tw-items-center">
+<div [formGroup]="form" class="tw-p-24-px tw-flex tw-flex-row tw-justify-between tw-items-center">
```
Comment (ashwin1111) [inline src/app/shared/components/configuration/configuration-select-field/configuration-select-field.component.html:1]: please test intacct and other places which uses this code

File: src/app/shared/components/configuration/configuration-select-field/configuration-select-field.component.html [hunk 17]
```diff
@@ -16,33 +16,59 @@ <h5 *ngIf="formControllerName==='cccExportType'" class="!tw-text-faded-text-colo
...
-                        {{option.label}}
-                    </div>
-                </ng-template>
-                <ng-template pTemplate="selectedItem" let-item>
-                    <div class="tw-flex tw-justify-between tw-items-center">
-                        <span>
-                            {{ item.label }}
-                        </span>
-                        <i class="pi pi-times" (click)="removeFilter(form.controls[formControllerName])"></i>
+        <div *ngIf="options && !destinationAttributes && !simpleOptions" class="tw-flex tw-justify-end">
```
Comment (ashwin1111) [inline src/app/shared/components/configuration/configuration-select-field/configuration-select-field.component.html:23]: what's changed here, diff looks pretty long

File: src/app/shared/components/configuration/configuration-toggle-field/configuration-toggle-field.component.html [hunk 18]
```diff
@@ -1,6 +1,9 @@
 <div [formGroup]="form" class="tw-p-24-px tw-flex tw-justify-between tw-items-center">
-    <div class="tw-w-3/4">
-      <div >
+    <div class="tw-w-3/4 tw-flex tw-items-start">
+      <div *ngIf="iconPath">
+        <svg-icon-sprite [src]="iconPath" width="20px" height="20px" class="tw-text-mandatory-field-color"></svg-icon-sprite>
+      </div>
+      <div class="tw-pl-16-px">
```
Comment (ashwin1111) [inline src/app/shared/components/configuration/configuration-toggle-field/configuration-toggle-field.component.html:6]: please test intacct and other places which uses this code

File: src/app/shared/components/configuration/email-multi-select-field/email-multi-select-field.component.html [hunk 19]
```diff
@@ -1,6 +1,6 @@
-<div [formGroup]="form" class="tw-p-24-px tw-flex tw-flex-row tw-space-between">
```
Comment (ashwin1111) [inline src/app/shared/components/configuration/email-multi-select-field/email-multi-select-field.component.html:1]: please test intacct and other places which uses this code

File: src/app/core/models/sage300/sage300-configuration/sage300-advanced-settings.mode.ts (hunk 4 above)
Comment (DhaaraniCIT) [reply src/app/core/models/sage300/sage300-configuration/sage300-advanced-settings.mode.ts:62]: just for now

File: src/app/shared/components/configuration/configuration-select-field/configuration-select-field.component.html (hunk 17 above)
Comment (DhaaraniCIT) [reply src/app/shared/components/configuration/configuration-select-field/configuration-select-field.component.html:23]: added a new ngIf for this
Screenshot 2023-11-15 at 11 55 37 PM

File: src/app/shared/components/configuration/email-multi-select-field/email-multi-select-field.component.html (hunk 19 above)
Comment (DhaaraniCIT) [reply src/app/shared/components/configuration/email-multi-select-field/email-multi-select-field.component.html:1]: no changes in intacct

File: src/app/shared/components/configuration/configuration-toggle-field/configuration-toggle-field.component.html (hunk 18 above)
Comment (DhaaraniCIT) [reply src/app/shared/components/configuration/configuration-toggle-field/configuration-toggle-field.component.html]: no changes in other apps

File: src/app/shared/components/configuration/configuration-select-field/configuration-select-field.component.html (hunk 16 above)
Comment (DhaaraniCIT) [reply src/app/shared/components/configuration/configuration-select-field/configuration-select-field.component.html:1]: no changes in other apps

File: src/app/shared/components/configuration/configuration-multi-select/configuration-multi-select.component.html (hunk 15 above)
Comment (DhaaraniCIT) [reply src/app/shared/components/configuration/configuration-multi-select/configuration-multi-select.component.html:1]: no changes in other apps

File: src/app/core/models/sage300/sage300-configuration/sage300-advanced-settings.model.ts [hunk 20]
```diff
@@ -0,0 +1,83 @@
...
+        };
+      }
+
+    static mapAPIResponseToFormGroup(advancedSettings: Sage300AdvancedSettingGet | null, isSkipExportEnabled: boolean): FormGroup {
+        return new FormGroup({
+            autoCreateMerchantDestinationEntity: new FormControl(advancedSettings?.auto_create_merchant_destination_entity ? true : false),
+            syncSage300ToFylePayments: new FormControl(advancedSettings?.sync_sage_300_to_fyle_payments ? true : false),
+            autoCreateDestinationEntity: new FormControl(advancedSettings?.auto_create_destination_entity ? true : false),
+            memoStructure: new FormControl(advancedSettings?.memo_structure ? advancedSettings?.memo_structure : null ),
+            defaultJobName: new FormControl(advancedSettings?.default_job_name ? advancedSettings?.default_job_name : null ),
```
Comment (ashwin1111) [inline src/app/core/models/sage300/sage300-configuration/sage300-advanced-settings.model.ts:57]: needs fixing, check intacct

File: src/app/core/models/sage300/sage300-configuration/sage300-advanced-settings.mode.ts (hunk 4 above)
Comment (ashwin1111) [reply src/app/core/models/sage300/sage300-configuration/sage300-advanced-settings.mode.ts]: please use fixture wherever you want some desired state, don't hard code anything in business logic

File: src/app/shared/components/configuration/configuration-select-field/configuration-select-field.component.html [hunk 21]
```diff
@@ -16,33 +16,59 @@ <h5 *ngIf="formControllerName==='cccExportType'" class="!tw-text-faded-text-colo
...
+            <div *ngIf="formControllerName !== 'dayOfMonth' && formControllerName !== 'dayOfWeek' && formControllerName !== 'timeOfDay' && !destinationAttributes">
```
Comment (DhaaraniCIT) [inline src/app/shared/components/configuration/configuration-select-field/configuration-select-field.component.html:24]: this is the start of the block mentioned @ashwin1111

File: src/app/shared/components/configuration/configuration-select-field/configuration-select-field.component.html [hunk 22]
```diff
@@ -16,33 +16,59 @@ <h5 *ngIf="formControllerName==='cccExportType'" class="!tw-text-faded-text-colo
...
+                <p-dropdown *ngIf="formControllerName !== 'scheduleAutoExportFrequency'" appendTo="body" [options]="options" class="!tw-w-300-px" 
+                    [formControlName]="formControllerName" [placeholder]="placeholder"
+                    [ngClass]="[form.controls[formControllerName].invalid && isFieldMandatory && form.controls[formControllerName].touched ? 'error-box' : 'normal-box']"
+                    [ngClass]="{'selectedItem': form.controls[formControllerName].value, 'filterDropdown': !form.controls[formControllerName].value}"
```
Comment (ashwin1111) [inline src/app/shared/components/configuration/configuration-select-field/configuration-select-field.component.html:28]: why ng class is repeated twice?

File: src/app/shared/components/configuration/configuration-select-field/configuration-select-field.component.html [hunk 23]
```diff
@@ -16,33 +16,59 @@ <h5 *ngIf="formControllerName==='cccExportType'" class="!tw-text-faded-text-colo
...
+                            </span>
+                            <i class="pi pi-times" (click)="removeFilter(form.controls[formControllerName])"></i>
+                        </div>
+                    </ng-template>
+                </p-dropdown>
+                <div class="tw-flex tw-justify-end" *ngIf="formControllerName === 'scheduleAutoExportFrequency'">
+                    <p-dropdown appendTo="body" [options]="options" class="dropdown-size-schedule" 
+                        [formControlName]="formControllerName" [placeholder]="placeholder"
+                        [ngClass]="[form.controls[formControllerName].invalid && isFieldMandatory && form.controls[formControllerName].touched ? 'error-box' : 'normal-box']"
+                        [ngClass]="{'selectedItem': form.controls[formControllerName].value, 'filterDropdown': !form.controls[formControllerName].value}"
```
Comment (ashwin1111) [inline src/app/shared/components/configuration/configuration-select-field/configuration-select-field.component.html:50]: why ng class is repeated twice?

File: src/app/core/models/sage300/sage300-configuration/sage300-advanced-settings.model.ts [hunk 24]
```diff
@@ -15,11 +16,10 @@ export type Sage300AdvancedSetting = {
...
     memo_structure: string[],
     default_job_name: string,
-    default_job_id: number
+    default_job_id: string
     schedule_enabled: boolean,
-    emails_selected: EmailOption[],
-    emails_added: EmailOption[],
-    auto_create_vendor: boolean
+    auto_create_vendor: boolean,
+    intervel_hours: number
```
Comment (ashwin1111) [inline src/app/core/models/sage300/sage300-configuration/sage300-advanced-settings.model.ts]: interval_hours, typo

File: src/app/integrations/sage300/sage300-shared/sage300-advanced-settings/sage300-advanced-settings.component.ts [hunk 25]
```diff
@@ -220,14 +222,15 @@ export class Sage300AdvancedSettingsComponent implements OnInit {
       this.advancedSettingsService.getAdvancedSettings().pipe(catchError(() => of(null))),
       this.exportSettingsService.getSage300ExportSettings(),
       this.advancedSettingsService.getExpenseFilter(),
-      this.advancedSettingsService.getExpenseFilelds()
-    ]).subscribe(([sage300AdvancedSettingResponse, exportSettingsResponse, expenseFiltersGet, ExpenseFilterCondition]) => {
+      this.advancedSettingsService.getExpenseFilelds(),
+      this.mappingService.getDestinationAttributes([Sage300Field.ACCOUNT], AppNameInService.SAGE300)
```
Comment (ashwin1111) [inline src/app/integrations/sage300/sage300-shared/sage300-advanced-settings/sage300-advanced-settings.component.ts]: type is wrong, it should be JOB

File: src/app/shared/components/configuration/configuration-select-field/configuration-select-field.component.ts [hunk 26]
```diff
@@ -52,6 +52,8 @@ export class ConfigurationSelectFieldComponent implements OnInit {
 
   @Input() exportTypeIconPathArray: any;
 
+  @Input() isDefaultFields: boolean = false;
```
Comment (ashwin1111) [inline src/app/shared/components/configuration/configuration-select-field/configuration-select-field.component.ts:55]: why do we need it separately?

File: src/app/shared/components/configuration/configuration-select-field/configuration-select-field.component.html [hunk 27]
```diff
@@ -1,109 +1,165 @@
-<div [formGroup]="form" class="tw-p-24-px tw-flex tw-flex-row tw-space-between tw-items-center">
```
Comment (ashwin1111) [inline src/app/shared/components/configuration/configuration-select-field/configuration-select-field.component.html:1]: revert this to master version, please don't format the file, it's not possible to review things. Incase if something is changed in this file, please point me to it

File: src/app/shared/components/configuration/configuration-select-field/configuration-select-field.component.ts (hunk 26 above)
Comment (DhaaraniCIT) [reply src/app/shared/components/configuration/configuration-select-field/configuration-select-field.component.ts:55]: this is for the default fields, in advanced setting, putting all select fields
Screenshot 2023-11-21 at 11 12 18 AM

**Review by ashwin1111 — APPROVED** [review]

**Pull request merged**